Add: global notice handlers for account recovery error actions

southp opened the pull request:
## Summary
This PR is a following act of the migrating the security checkup store from `flux` to `redux`, initiating by #9109 . It is based on #9111 , thus should only be labeled as "needs review" after it.

It adds all the handlers to the global notice middleware for the error actions from the `account-recovery` redux store. Namely,

* `ACCOUNT_RECOVERY_FETCH_FAILED`
* `ACCOUNT_RECOVERY_PHONE_UPDATE_FAILED`
* `ACCOUNT_RECOVERY_EMAIL_UPDATE_FAILED`
* `ACCOUNT_RECOVERY_PHONE_DELETE_FAILED`
* `ACCOUNT_RECOVERY_EMAIL_DELETE_FAILED`

To make `state/notices/middleware.js` less clutter, it also introduces the handlers in its own subdirectory, and move `dispatchSuccess` and `dispatchError` utility functions into its own module, `utils.js`.

## Test Plan
If your redux devtool could dispatch actions properly, the fastest way would be dispatching those actions through it, and see if the global notices pop up properly. In my case, my devtool doesn't work, and thus I added the following line in the `reduxStoreReady( reduxStore )` function:
```
window.testDispatch = ( action ) => reduxStore.dispatch( action );
```
and then call this newly-added `testDispatch` function from the debug console to see them in action.

southp committed:
Fix a typo: Suported => Supported

southp committed:
Add error handlers for the error actions of the `account recovery`
state.

southp committed:
Move the error handlers related to `account-recovery` error actions to
its own subdirectory to make the middleware part less clutter.

- `dispatchSuccess` and `dispatchError` are also moved to its own
  module, `utils.js`, to prevent cycle dependency.

southp committed:
fix: `dispatchSuccess` and `dispatchError` are now in `utils.js` instead of `middleware.js`

southp committed:
Rewrite the notice handlers based on the changes to account recovery
state actions in the parent branch.

jordwest commented on client/state/account-recovery/settings/actions.js:
```diff
@@ -0,0 +1,164 @@
+		).catch( ( { status, message } ) =>
+			dispatch( deleteAccountRecoveryEmailFailed( { status, message } ) )
+		);
+};
```
There's a lot of repetition in this file. Generally we have one action creator per action type (eg, pass in the target as a param), so that could reduce some of this duplication.

Whether it's necessary is up to you, but it might also be possible to reduce duplication for the thunk bodies using higher order functions if that seems like a reasonable approach.

jordwest replied:
Whoops, got ahead of myself here and reviewed the wrong PR. Moved this comment to: https://github.com/Automattic/wp-calypso/pull/9111/files#r87335896

jordwest commented on client/state/notices/account-recovery/index.js:
```diff
@@ -0,0 +1,48 @@
+		default:
+			return errorNotice( translate( 'An error occurred while deleting your account recovery options.' ) );
+	}
+};
```
`getDeleteErrorNotice` could instead just return the text:

```js
const getDeleteErrorText = ( target ) => {
	switch ( target ) {
		case 'phone':
			return translate( 'An error occurred while deleting your account recovery phone number.' );
		case 'email':
			return translate( 'An error occurred while deleting your account recovery email.' );
		default:
			return translate( 'An error occurred while deleting your account recovery options.' );
	}
};

// ...

[ ACCOUNT_RECOVERY_SETTINGS_DELETE_FAILED ]: ( dispatch, { target } ) => dispatch( errorNotice( getDeleteErrorText( target ) ) ),

```

southp replied:
Fixed by ef011a61fb9ba51e1af71aa062d22e4b5f36dcc5

southp committed:
`getDeleteErrorNotice()` is renamed as `getDeleteErrorMessage()`, and is
simplified as returning String only. So does `getUpdateErrorNotice()`.

southp commented:
Now that #9111 is merged, this is ready for review.
cc @jordwest @omarjackman

omarjackman commented on client/state/notices/account-recovery/index.js:
```diff
@@ -0,0 +1,48 @@
+
+const handlers = {
+	[ ACCOUNT_RECOVERY_SETTINGS_FETCH_FAILED ]: dispatchError(
+		translate( 'An error occurred while fetching for your account recovery settings.' )
```
This should probably be "An error occurred while fetching ~~for~~ your account recovery settings."

southp replied:
Fixed by ce45f890d61fdcba87a17b811c078ef31313e000

omarjackman commented on client/state/notices/account-recovery/index.js:
```diff
@@ -0,0 +1,48 @@
+	[ ACCOUNT_RECOVERY_SETTINGS_FETCH_FAILED ]: dispatchError(
+		translate( 'An error occurred while fetching for your account recovery settings.' )
+	),
+	[ ACCOUNT_RECOVERY_SETTINGS_UPDATE_FAILED ]: ( dispatch, { target } ) => dispatch( errorNotice( getUpdateErrorMessage( target ) ) ),
```
This is just a question/observation more than it is a criticism :)...

Why the use of `errorNotice` here but not in the handler for `ACCOUNT_RECOVERY_SETTINGS_FETCH_FAILED`. I notice that you removed it in https://github.com/Automattic/wp-calypso/commit/ef011a61fb9ba51e1af71aa062d22e4b5f36dcc5

southp replied:
It is because `dispatchError` is a utility function very suitable for the case of `ACCOUNT_RECOVERY_SETTINGS_FETCH_FAILED`, which no extra action properties need to be extracted. However, in case of `ACCOUNT_RECOVERY_SETTINGS_UPDATE_FAILED` and `ACCOUNT_RECOVERY_SETTINGS_DELETE_FAILED`, it would become:
```
( dispatch, { target } ) => dispatchError( getUpdateErrorMessage( target ) )( dispatch )
```
which I think that it is not clearer than just using `errorNotice`.

Lastly, what happens in ef011a61fb9ba51e1af71aa062d22e4b5f36dcc5 is to address [Jordan's comment](https://github.com/Automattic/wp-calypso/pull/9173#discussion_r87335937). I took it since I think that would make the function simpler. 

Does that answer your question? :)

southp committed:
Fix a grammar error: "fetching for" => "fetching"

southp commented:
@jordwest @omarjackman If there is some spare time, could you help me revisit to check if there is any concerns left? Thanks :)

jordwest commented on client/state/notices/middleware.js:
```diff
@@ -91,7 +82,8 @@ export function onPostSaveSuccess( dispatch, action ) {
 export const handlers = {
-	[ ACCOUNT_RECOVERY_FETCH_FAILED ]: dispatchError( translate( 'An error occurred while fetching for your account recovery settings.' ) ),
+	...accountRecoveryNoticeHandlers,
+
```
Not sure about others, but I'd prefer to see the actions listed here, as it's not otherwise clear which actions are being handled by this middleware without tracing through to the import.

I think it's a good move to keep the notices themselves in a separate module though, as this file is going to keep growing.

southp replied:
Oh, and 4b1373f29f4deda9732c8da7421cdf057aac1926. I forgot to remove the import statement in the previous commit :)

jordwest reviewed (commented):
I've left a small suggestion, but otherwise it looks like a good improvement.

southp committed:
Leave only notice handler functions in `notices/account-recovery`
module, and explicitly add these handlers in `notices/middleware.js`.
This way, we can have one single general view of what actions are
handled here.

southp committed:
Remove an empty import statement.

southp commented:
@jordwest I forgot to mention that your comments are addressed :)

jordwest commented:
LGTM! 🚢
